golem-worker-service-base/src/parser/expr/expr_parser.rs

```diff
@@ -1,495 +1,299 @@
-                                Box::new(prev),
-                                Box::new(new),
-                            ))
-                        },
-                    )?;
-
-                    go(tokenizer, context, new_expr)
-                }
+                let new_expr =
+                    math_op::create_binary_op(&mut tokenizer, left_op, |left, right| {
+                        Expr::GreaterThan(left, right)
+                    })?;
 
-                Token::Dot => match prev_expression {
-                    InternalExprResult::Complete(expr) => {
-                        let expr = selection::get_select_field(tokenizer, expr)?;
-                        go(tokenizer, context, InternalExprResult::complete(expr))
-                    }
+                previous_expression.build(new_expr);
+            }
 
-                    _ => Err(ParseError::message(
-                        "Invalid token field. Make sure expression format is correct",
-                    )),
-                },
-
-                Token::LSquare => match prev_expression {
-                    InternalExprResult::Complete(prev_expr) => {
-                        let new_expr = selection::get_select_index(tokenizer, &prev_expr)?;
-                        go(tokenizer, context, InternalExprResult::complete(new_expr))
-                    }
-                    _ => {
-                        let expr = sequence::create_sequence(tokenizer)?;
-                        go(tokenizer, context, prev_expression.apply_with(expr))
-                    }
-                },
-
-                Token::MultiChar(MultiCharTokens::If) => {
-                    // We expect to form Expr::Cond given three unknown variables
-                    let future_expr = InternalExprResult::incomplete(
-                        InCompleteExpressionContext::If,
-                        move |first_result| {
-                            let first_result: Rc<Expr> = Rc::new(first_result);
-                            InternalExprResult::incomplete(
-                                InCompleteExpressionContext::Then,
-                                move |second_result| {
-                                    let first_result: Rc<Expr> = Rc::clone(&first_result);
-                                    InternalExprResult::incomplete(
-                                        InCompleteExpressionContext::Else,
-                                        move |else_result| {
-                                            let first_result: Expr =
-                                                (*Rc::clone(&first_result)).clone();
-                                            InternalExprResult::complete(Expr::Cond(
-                                                Box::new(first_result),
-                                                Box::new(second_result.clone()),
-                                                Box::new(else_result),
```

Comment:
Forming higher order functions was logiccally ok, but it just giving me lot of pain with Rust